src/asound/bluealsa-pcm.c

```diff
@@ -666,47 +666,84 @@ static int bluealsa_drain(snd_pcm_ioplug_t *io) {
 	struct bluealsa_pcm *pcm = io->private_data;
 	debug2("Draining");
 
-	if (!pcm->connected) {
-		snd_pcm_ioplug_set_state(io, SND_PCM_STATE_DISCONNECTED);
-		return -ENODEV;
-	}
-
 	/* A bug in the ioplug drain implementation means that snd_pcm_drain()
 	 * always either finishes in state SND_PCM_STATE_SETUP or returns an error.
 	 * It is not possible to finish in state SND_PCM_STATE_DRAINING and return
 	 * success; therefore is is impossible to correctly implement capture
 	 * drain logic. So for capture PCMs we do nothing and return success. */
+	if (io->stream == SND_PCM_STREAM_CAPTURE)
+		return 0;
 
-	if (io->stream == SND_PCM_STREAM_PLAYBACK) {
-
-		/* We must ensure that all remaining frames in the ring buffer are
-		 * flushed to the FIFO by the I/O thread. It is possible that the
-		 * client has called snd_pcm_drain() without the start_threshold
-		 * having been reached, or while paused, so we must first ensure that
-		 * the IO thread is running. */
-		if (bluealsa_start(io) < 0)
-			return 0;
-
-		/* Block until the drain is complete. */
-		struct pollfd pfd = { pcm->event_fd, POLLIN, 0 };
-		while (bluealsa_pointer(io) >= 0 && io->state == SND_PCM_STATE_DRAINING) {
-			if (poll(&pfd, 1, -1) == -1) {
-				if (errno == EINTR)
+	/* We must ensure that all remaining frames in the ring buffer are flushed
+	 * to the FIFO by the I/O thread. It is possible that the client has called
+	 * snd_pcm_drain() without the start_threshold having been reached, or
+	 * while paused, so we must first ensure that the IO thread is running. */
+	if (bluealsa_start(io) < 0)
```

Comment:
In case of `bluealsa_start()` failure the error is not returned intentionally? If not, maybe it could be a good idea to return error from drain function as well?

Reply:
The current code simply copies the ioplug built-in drain behaviour by returning zero for any error - ioplug then always calls snd_pcm_ioplug_drop().

If we return an error code, then ioplug will return that code to the client, without changing the PCM state or dropping the stream. So we would need to determine and set the correct state for each error, and possibly also drop the stream if that is appropriate.